activation/nipost.go

```diff
@@ -365,98 +394,116 @@ func (nb *NIPostBuilder) submitPoetChallenge(
 
 	logger.Debug("submitting challenge to poet proving service")
 
-	round, err := client.Submit(ctx, deadline, prefix, challenge, signature, nodeID)
+	registration := nipost.PoETRegistration{
+		NodeId:        nodeID,
+		ChallengeHash: types.Hash32(challenge),
+		Address:       client.Address(),
+	}
+
+	round, err := client.Submit(ctx, fetchProofDeadline, prefix, challenge, signature, nodeID)
 	if err != nil {
+		registration.RoundEnd = poetRoundEnd
+
+		if err := nipost.AddPoetRegistration(nb.localDB, registration); err != nil {
+			nb.logger.Warn("failed to save failed poet registration into db",
+				zap.Error(err),
+				log.ZShortStringer("smesherID", nodeID))
+		}
 		return nipost.PoETRegistration{},
 			&PoetSvcUnstableError{msg: "failed to submit challenge to poet service", source: err}
 	}
+
 	logger.Info("challenge submitted to poet proving service", zap.String("round", round.ID))
 
-	registration := nipost.PoETRegistration{
-		ChallengeHash: types.Hash32(challenge),
-		Address:       client.Address(),
-		RoundID:       round.ID,
-		RoundEnd:      round.End,
-	}
+	registration.RoundEnd = round.End
+	registration.RoundID = round.ID
 
-	if err := nipost.AddPoetRegistration(nb.localDB, nodeID, registration); err != nil {
+	err = nipost.AddPoetRegistration(nb.localDB, registration)
+	switch {
+	case err != nil && errors.Is(err, sql.ErrObjectExists):
+		if err := nipost.UpdatePoetRegistration(nb.localDB, registration); err != nil {
+			return nipost.PoETRegistration{}, err
+		}
+
+	case err != nil:
 		return nipost.PoETRegistration{}, err
 	}
-
-	return registration, err
+	return registration, nil
```

Comment:
Instead of trying to add and then updating when it fails we could change the `INSERT` in `AddPoetRegistration` to an `UPSERT`. Opinions @poszu ?

Reply:
SGTM :+1: